test/test-queue-foreach-delete.c

```diff
@@ -0,0 +1,211 @@
+      uv_close((uv_handle_t*)&(name)[0], NULL);                               \
+      uv_close((uv_handle_t*)&(name)[1], NULL);                               \
+    }                                                                         \
+    name##_cb_calls[0]++;                                                     \
+  }                                                                           \
+                                                                              \
+  static const uv_##name##_cb name##_cbs[] = {                                \
+    (uv_##name##_cb)name##0_cb,                                               \
+    (uv_##name##_cb)name##1_cb,                                               \
+    (uv_##name##_cb)name##2_cb,                                               \
+  };
+
+#define INIT_AND_START(name, loop)                                            \
+  do {                                                                        \
+    size_t i;                                                                 \
+    for (i = 0; i < ARRAY_SIZE(name); i++) {                                  \
+      int r;                                                                  \
+      r = uv_##name##_init((loop), &(name)[i]);                               \
+      ASSERT(r == 0);                                                         \
+                                                                              \
+      r = uv_##name##_start(&(name)[i], name##_cbs[i]);                       \
+      ASSERT(r == 0);                                                         \
+    }                                                                         \
+  } while (0)
+
+#define END_ASSERTS(name)                                                     \
+  do {                                                                        \
+    ASSERT(name##_cb_calls[0] == 1);                                          \
+    ASSERT(name##_cb_calls[1] == 0);                                          \
+    ASSERT(name##_cb_calls[2] == 1);                                          \
+  } while (0)
+
+DEFINE_GLOBALS_AND_CBS(idle)
+DEFINE_GLOBALS_AND_CBS(prepare)
+DEFINE_GLOBALS_AND_CBS(check)
+DEFINE_GLOBALS_AND_CBS(async)
+DEFINE_GLOBALS_AND_CBS(fs_event)
+
+static const char watched_dir[] = ".";
+
+static void init_asyncs(uv_loop_t* loop) {
+  size_t i;
+  for (i = 0; i < ARRAY_SIZE(async); i++) {
+    int r;
+    r = uv_async_init(loop, &async[i], async_cbs[i]);
+    ASSERT(r == 0);
+  }
+}
+
+static void init_and_start_fs_events(uv_loop_t* loop) {
+  size_t i;
+  for (i = 0; i < ARRAY_SIZE(fs_event); i++) {
+    int r;
+    r = uv_fs_event_init(loop, &fs_event[i]);
+    ASSERT(r == 0);
+
+    r = uv_fs_event_start(&fs_event[i], (uv_fs_event_cb)fs_event_cbs[i], watched_dir, 0);
+    ASSERT(r == 0);
+  }
+}
+
+
+#ifdef _WIN32
+static int mytouch(const char *filename) {
+  return _utime(filename, NULL);
```

Comment:
> why not use uv_fs_utime?

Thank you for the suggestion!
I think I decided to go with synchronous utime() call for simplicity.
I'll try to use uv_fs_utime().